src/ghci/mod.rs

```diff
@@ -98,6 +102,10 @@ pub struct GhciOpts {
     pub restart_globs: GlobMatcher,
     /// Reload the `ghci` session when paths matching these globs are changed.
     pub reload_globs: GlobMatcher,
+    /// Where to write what `ghci` emits to `stdout`. Inherits parent's `stdout` by default.
+    pub stdout_writer: Box<dyn GhciWrite>,
+    /// Where to write what `ghci` emits to `stderr`. Inherits parent's `stderr` by default.
+    pub stderr_writer: Box<dyn GhciWrite>,
```

Comment:
Maybe instead of type parameters or `Box<dyn>`, we could have an enum with all the types we expect:

```rust
enum GhciStdout {
  Stdout(...),
  Tui(...),
}
```

And then `impl` `AsyncWrite` and whatnot on that?

I'd really love to avoid `Box<dyn Arc<Mutex<T>>>`, but I suspect in practice the overhead won't be too bad if we can't find something nicer.

Reply:
> Maybe instead of type parameters or Box<dyn>, we could have an enum with all the types we expect

Good idea! That'd remove the `Box` in all cases, and the `Arc` + `Mutex` in the `Stdout`/`Stderr` cases (`clone` would just create another via `tokio::io::std{out,err}()`).

> I suspect in practice the overhead won't be too bad

I agree. I think it's more objectionable in theory than in practice. I think the terminal and the user's brain have lower throughput than this lol.

Reply:
I implemented this in https://github.com/MercuryTechnologies/ghciwatch/pull/175/commits/145f5ec71e0ebebd850f544c1816c1dd528ef87b.

I think the idea you mentioned in Slack - keeping `AsyncWrite`rs out of `GhciOpts`, opting to create them during startup or restart and send them over a channel - is a good one, but I'd rather explore that in a future PR if that's okay.